src/kinship_pruning.jl

```diff
@@ -0,0 +1,140 @@
+- `:top_down` runs top-down greedy method, iteratively eliminating sample with the highest degree.
+- `:gcta` runs the method implemented in GCTA.
+- `:plink` runs the method implemented in PLINK.
+
+In general, `:bottom_up` greedy method performs the best (keeping the hightest number of subjects) and has a guaranteed approximamtion ratio.
+With `min_samples` defined, `:top_down` greedy method keeps the least number of off-diagonal values.
+"""
+function kinship_pruning(grm::Matrix; method=:bottom_up, cutoff=0.25, min_samples=0)
+    if method == :bottom_up
+        r = _kinship_pruning_bottomup(grm; cutoff=cutoff, min_samples=min_samples)
+    elseif method == :top_down
+        r = _kinship_pruning_topdown(grm; cutoff=cutoff, min_samples=min_samples)
+    elseif method == :gcta
+        r = _kinship_pruning_gcta(grm; cutoff=cutoff)
+    elseif method == :plink
+        r = _kinship_pruning_plink(grm; cutoff=cutoff, min_samples=min_samples)
+    else
+        @error "Invalid value of argument method. Valid values are: :bottom_up, :top_down, :gcta, and :plink."
+    end
+    r
+end
+
+function _find_neighbors(i::Int, grm::Matrix, r::BitVector, degrees::Vector{Int}; cutoff=0.25)
+    m = size(grm, 1)
+    neighbors = Array{Int}(undef, degrees[i])
+    n_neighbors = 0
+    for j in 1:m
+        if r[j] && i != j && grm[i, j] > cutoff
+            n_neighbors += 1
+            neighbors[n_neighbors] = j
+        end
+    end
+    @assert n_neighbors == degrees[i]
+    neighbors
+end
+
+function _count_degrees(grm::Matrix; cutoff=0.25)
+    m = size(grm, 1)
+    degrees = zeros(Int, m)
+    for i in 1:m
+        for j in 1:m
+            if j == i; continue; end
+            if grm[i, j] > cutoff
+                degrees[i] += 1
+            end
+        end
+    end
+    degrees
+end
+
+function _eliminate_node!(i::Int, grm::Matrix, r::BitVector, degrees::Vector{Int}; cutoff=0.25)
+    m = size(grm, 1)
+    r[i] = false
+    degrees[i] = 0
+    for j in 1:m
+        if r[j] && grm[i, j] > cutoff
+            degrees[j] -= 1
+        end
+    end
+    nothing 
+end
+
+function _kinship_pruning_gcta(grm::Matrix; cutoff=0.25)
+    m = size(grm, 1)
+    degrees = _count_degrees(grm; cutoff=cutoff)
+    
+    r = trues(m)
+    for i in 1:m
+        for j in (i+1):m
+            if j == i; continue; end
+            if grm[i, j] > cutoff
+                if degrees[i] > degrees[j] 
+                    r[i] = false
+                else
+                    r[j] = false
+                end
+            end
+        end
+    end 
+    r
+end
+
+function _kinship_pruning_plink(grm::Matrix; cutoff=0.25, min_samples=0)
+    m = size(grm, 1)
+    degrees = _count_degrees(grm; cutoff=cutoff) 
+    r = trues(m) 
+    while sum(degrees) > 0 && count(r) > min_samples
+        if any(degrees .== 1)
+            to_add = findfirst(x -> x == 1, degrees)
+            to_eliminate = _find_neighbors(to_add, grm, r, degrees; cutoff=cutoff)
```

Comment:
Is this better to have an in-place version of `_find_neighbors` for efficiency? So we don't need to allocate the result vector in each iteration. Just reuse the same vector.